### PR TITLE
[8.x] Add a basic deprecation warning that the JSON format is changing in v9 (#116330)

### PR DESCRIPTION
**Pull request by thecoop**
Backport #116330 to 8.17

### CODE REVIEW

**Comment by thecoop** [comment]
I'm not sure if this needs to be a critical (as things are definitely changing), or a warning, as the user can't do anything about it until they upgrade.

Do we need to include a link to more detailed docs (which don't really exist, apart from the breaking changelog?)

**Comment by DaveCTurner** [comment]
> the user can't do anything about it until they upgrade.

That seems bad to me. Certainly this can't be critical because critical means "must be addressed _before_ you upgrade". But could we have a way for users to opt-in to the new format ahead of time? Should we fall back to the old format in v9 if the user requests `RestApiVersion.V_8`?

**Comment by thecoop** [comment]
This is the behaviour as agreed in the BwC discussion...

**Comment by DaveCTurner** [comment]
Ok I see, I wasn't in that discussion 😄 

Can we tell folks to opt-in to the future behaviour now by removing `http.detailed_errors.enabled` from their node settings?

**Comment by DaveCTurner** [comment]
Sure but we've got to do _something_ here, we can't just expect folks to upgrade and suck it up if their clients can't handle the new format caused by the upgrade. Either they need to fix their clients before upgrading or else we need to preserve the v8 error format in v9.

**Comment by thecoop** [comment]
That's what the decision of the BwC committee was - there's very little use of this option, so any attempt at pro-active mitigation is disproportionate

**Comment by DaveCTurner** [comment]
> That's what the decision of the BwC committee was

That's not my reading of the decision. There was no mention of this change affecting clients that requested v8 compatibility in v9, and our default position for that case would be that such clients see no change in behaviour. Would you ask the BwC committee to clarify whether they intended to make this change in v8-compat mode too?

**Comment by rjernst** [comment]
I agree with Dave's concern. The point of rest api compatibility is to bridge these exact gaps, changes across majors. I think we should maintain the existing format when the v8 header is used.

**Comment by thecoop** [comment]
I've added BwC API outputs to https://github.com/elastic/elasticsearch/pull/90529

File: server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java [hunk 1]
```diff
@@ -61,6 +64,15 @@ protected AbstractRestChannel(RestRequest request, boolean detailedErrorsEnabled
         this.filterPath = request.param("filter_path", null);
         this.pretty = request.paramAsBoolean("pretty", false);
         this.human = request.paramAsBoolean("human", false);
+
+        if (detailedErrorsEnabled == false) {
+            deprecationLogger.critical(
+                DeprecationCategory.API,
+                "http_detailed_errors",
+                "The JSON format of non-detailed errors will change in Elasticsearch 9.0 to match the JSON structure"
+                    + " used for detailed errors. To keep using the existing format, use the V8 REST API."
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java:73]: We need the same deprecation warning in the v9 code too (when hitting this path in v8 compat mode).

I also don't think we can have a critical warning without any way for folks to resolve it before upgrading.

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java:73]: I've changed it to a warning, and will copy these changes to the v9 PR

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java:73]: This is now done

**Review by DaveCTurner — COMMENTED** [review]
LGTM except that I'd rather we followed the usual process of merging this into `main`, backporting it to `8.x`, and then merging the remnants of #90529 into `main`. As it is, it's overly hard to review the difference between this PR and #90529.

**Comment by thecoop** [comment]
Commit https://github.com/elastic/elasticsearch/pull/90529/commits/eafd2ea6b3245b037fc1320d81ab12267089ac25 is basically this PR cherry-picked onto main. It doesn't have the actual JSON structure change though.

**Comment by thecoop** [comment]
See https://github.com/elastic/elasticsearch/pull/116330

**Comment by thecoop** [comment]
@elasticmachine update branch

File: docs/changelog/114739.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+pr: 114739
+summary: Add a basic deprecation warning that the JSON format for non-detailed error responses is changing in v9
+area: Infra/REST API
+type: deprecation
+issues: [89387]
+deprecation:
+  title: The format of non-detailed error responses is changing in v9
+  area: REST API
+  details: |-
```
Comment (mosche) [inline docs/changelog/114739.yaml:9]: Add `To keep using the existing format in v9, use the V8 REST API.` here as well?

**Review by mosche — APPROVED** [review]
LGTM, just a small suggestion.

File: docs/changelog/114739.yaml (hunk 2 above)
Comment (thecoop) [reply docs/changelog/114739.yaml:9]: Good spot, changed the changelog

**Pull request merged**